Implement support for default values for optional expected keys

gee-forr opened the pull request:
This PR extends the `expects` macro so that a default value can be supplied. When an expected key's value is not supplied, and there is a default, it will be used instead, and a `LightService::ExpectedKeysNotInContextError` will be avoided.

I've implemented support for both static default values and dynamic defaults. Dynamic defaults accept a callable that takes the context as a param.

Full documentation on how to use this feature is included. Tests are also included.

This should also resolve #215. It does also obsolete https://github.com/adomokos/light-service/pull/216 unfortunately.

Hope you like it :)

gee-forr commented:
@adomokos - here you go - hope you like it. Let me know if there's anything I should change. The biggest changes are in the docs to be honest. I've only really made a few changes in `lib/light-service/actions.rb` at two stages:

1. When processing all expected keys - I store defaults for later use
2. Just prior to creating the context - I fill in missing expects that have defaults that should be used.

adomokos reviewed (commented):
This is looking really great @gee-forr, thank you for your work on it.
I had a couple of suggestions, let me know what you think.

adomokos commented:
@gee-forr - [this is](https://github.com/adomokos/light-service/pull/217) coming your way.
I am looking into reducing the complexity of the `executed` function.

gee-forr commented on lib/light-service/action.rb:
```diff
@@ -15,6 +15,13 @@ def self.included(base_class)
     module Macros
       def expects(*args)
+        if expect_key_having_default?(args)
+          available_defaults[args.first] = args.last[:default] \
```
If someone makes a mistake, i.e. `expects :foo, defalut: "bar"`, the default is ignored, but the key is still marked as expected. In this case, a standard `ExpectedKeysNotInContextError` exception will be raised.

It's better that we catch it at define time, and not execute time. I'll make it throw an exception.

gee-forr replied:
OK - this is done :)

gee-forr commented on lib/light-service/action.rb:
```diff
@@ -76,6 +93,22 @@ def all_keys
+
+        default.call(context)
+      end
+
```
I definitely think it could do with a separate class or module handling key setup as a first step, and then extracting that class into two `expects` and `promises` classes if it needs more refactoring.

Right now, the action module is doing the following things:

1. setting up expects and promises keys (incl. defaults when supplied)
2. executing an action
3. rolling back an action
4. handling callbacks
5. preparing a context
6. verifying key existence before an action

All tasks can be grouped into two categories:

1. Preparation - parsing keys, setting up context, verifying keys
2. Execution - execute, roll back, callbacks

I reckon refactoring the preparation code into a single responsibility will be a huge step forward without necessitating an overhaul. A good effort:result ratio in my opinion.

What do you think. I'd be keen to take a stab at this next month...

gee-forr commented on README.md:
```diff
@@ -529,56 +531,90 @@ These ideas are originally from Aspect Oriented Programming, read more about the
-  expects :baz
-  promises :bar
+  expects :message, :user
+  expects :allow_failure, default: ->(ctx) { !ctx[:user].admin? } # Admins must always get SMS'
```
thanks - was struggling to find an example that made sense as a dynamic value

gee-forr commented on lib/light-service/action.rb:
```diff
@@ -30,6 +37,14 @@ def promised_keys
 
+      def available_defaults
+        @available_defaults ||= {}
+      end
```
Agreed - there's a bunch that can be made private. I'll make everything private that can be. Will help with reasoning about the modules actual API for later refactor simpler too.

gee-forr replied:
All done.

gee-forr commented on lib/light-service/action.rb:
```diff
@@ -67,7 +82,9 @@ def rolled_back
 
       def create_action_context(context)
-        return context if context.is_a? LightService::Context
+        usable_defaults(context).each do |ctx_key, default|
```
`map` wouldnt be appropriate here as we're not returning a new enumerable, we're mutating the candidate context blob to apply defaults if they're required.

`usable_defaults(ctx)` takes the list of missing expected keys, and returns a trimmed hash of available defaults that contains only the missing keys that have a default.

It then loops through that hash and applies the default to the context.

Can you walk me through how this would work with a map?

gee-forr commented on lib/light-service/action.rb:
```diff
@@ -30,6 +37,14 @@ def promised_keys
+
+      def expect_key_having_default?(key)
+        key.size == 2 && key.last.is_a?(Hash)
+      end
```
All done :)

gee-forr commented:
@adomokos - ok - I've pulled in the latest changes to `main`, resolved the conflicts, and addressed your points. Let me know if there's anything else I need to work on.

Thanks for the high quality review.

metalelf0 commented:
Hey! This is really great and much better than my attempt. Congrats for getting it done! 👍🏻

gee-forr commented:
> Hey! This is really great and much better than my attempt. Congrats for getting it done! 👍🏻

Thanks for the kind words :) I hope it helps you with your particular use case :)

adomokos commented:
I'll try to get a release out this weekend.
